lib/auth/ec2_join.go

```diff
@@ -0,0 +1,755 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/credentials/stscreds"
+	"github.com/aws/aws-sdk-go-v2/feature/ec2/imds"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	ec2types "github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+	"go.mozilla.org/pkcs7"
+)
+
+type ec2Client interface {
+	DescribeInstances(ctx context.Context, params *ec2.DescribeInstancesInput, optFns ...func(*ec2.Options)) (*ec2.DescribeInstancesOutput, error)
+}
+
+type ec2ClientKey struct{}
+
+func ec2ClientFromContext(ctx context.Context) (ec2Client, bool) {
+	ec2Client, ok := ctx.Value(ec2ClientKey{}).(ec2Client)
+	return ec2Client, ok
+}
+
+// ec2ClientFromConfig returns a new ec2 client from the given aws config, or
+// may load the client from the passed context if one has been set (for tests).
+func ec2ClientFromConfig(ctx context.Context, cfg aws.Config) ec2Client {
+	ec2Client, ok := ec2ClientFromContext(ctx)
+	if ok {
+		return ec2Client
+	}
+	return ec2.NewFromConfig(cfg)
+}
+
+// NodeIDFromIID returns the node ID that must be used for nodes joining with
+// the given Instance Identity Document.
+func NodeIDFromIID(iid *imds.InstanceIdentityDocument) string {
+	return iid.AccountID + "-" + iid.InstanceID
+}
+
+// checkEC2AllowRules checks that the iid matches at least one of the allow
+// rules of the given token.
+func checkEC2AllowRules(ctx context.Context, iid *imds.InstanceIdentityDocument, token types.ProvisionToken) error {
+	allowRules := token.GetAllowRules()
+	for _, rule := range allowRules {
+		if len(rule.AWSAccount) > 0 {
+			if rule.AWSAccount != iid.AccountID {
+				continue
+			}
+		}
+		if len(rule.AWSRegions) > 0 {
+			matchingRegion := false
+			for _, region := range rule.AWSRegions {
+				if region == iid.Region {
+					matchingRegion = true
+					break
+				}
+			}
+			if !matchingRegion {
+				continue
+			}
+		}
+		// iid matches this allow rule. Check if it is running.
+		return trace.Wrap(checkInstanceRunning(ctx, iid.InstanceID, iid.Region, rule.AWSRole))
```

Comment:
Do we care if the node stops running between now and the point of registration? I don't _think_ we do, but best to check.

Reply:
I don't think we would care - the point of the check is to make sure the IID isn't completely bogus and this is a real EC2 instance that actually was running at roughly the time the register request was sent. If a legitimate EC2 instance dies in the middle of registration, that's fine